src/pressure/VCycleMG.cpp

```diff
@@ -248,171 +215,88 @@ void VCycleMG::pressure(Field &out, Field const &b, real t, bool sync) {
-            size_t start_b = boundary->get_boundary_list_level_joined_start(i + 1);
-            size_t end_b = boundary->get_boundary_list_level_joined_end(i + 1) + 1;
-            // inner
-#pragma acc kernels present(err1ip, d_iList[start_i:(end_i-start_i)]) async
-#pragma acc loop independent
-            for (size_t j = start_i; j < end_i; ++j) {
-                const size_t idx = d_iList[j];
-                d_err1ip[idx] = 0.0;
-            }
+                // restrict
+                Restrict(*field_residuum1_level_plus_1, *field_residuum0_level, level, sync);
+                boundary->apply_boundary(data_residuum1_level_plus_1, level + 1, type_r1, sync); // for res only Dirichlet BC
 
-            // boundary
-#pragma acc kernels present(err1ip, d_bList[start_b:(end_b-start_b)]) async
-#pragma acc loop independent
-            for (size_t j = start_b; j < end_b; ++j) {
-                const size_t idx = d_bList[j];
-                d_err1ip[idx] = 0.0;
+                // set err to zero at next level
+                field_error1_level_plus_1->set_value(0);
             }
         }
-    }
 
-//===================== levels going up ====================//
-    for (size_t i = max_level; i > 0; --i) {
-        Field &f_err0i = *err0[i];
-        Field &f_err1i = *error1[i];
-        Field &f_err1im = *error1[i - 1];
-        Field &f_mg_tmpim = *mg_temporal_solution[i - 1];
-        Field &f_res1im = *residuum1[i - 1];
+        //===================== levels going up ====================//
+        for (size_t level = m_levels; level > 0; --level) {
+            Field *field_error0_level = *(m_error0 + level - 1);
+            Field *field_error1_level = *(m_error1 + level);
+            Field *field_error1_level_minus_1 = *(m_error1 + level - 1);
+            Field *field_mg_temporal_solution_level_minus_1 = *(m_mg_temporal_solution + level - 1);
+            Field *field_residuum1_level_minus_1 = *(m_residuum1 + level - 1);
 
-        FieldType type_e0 = f_err0i.get_type();
+            real *data_error0_level = field_error0_level->data;
 
-        // inner start/ end index of level i - 1
-        size_t start_i = boundary->get_inner_list_level_joined_start(i - 1);
-        size_t end_i = boundary->get_inner_list_level_joined_end(i - 1) + 1;
-        // boundary start/ end index of level i - 1
-        size_t start_b = boundary->get_boundary_list_level_joined_start(i - 1);
-        size_t end_b = boundary->get_boundary_list_level_joined_end(i - 1) + 1;
+            FieldType type_e0 = field_error0_level->get_type();
+
+#pragma acc data present(error0_level, error1_level, error1_level_minus_1, mg_temporal_solution_level_minus_1, residuum1_level_minus_1, out)
+            {
+                // prolongate
+                Prolongate(*field_error0_level, *field_error1_level, level, sync);
+                boundary->apply_boundary(data_error0_level, level - 1, type_e0, sync); // for m_error0 only Dirichlet BC
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247